File: Test/src/test/scala/test/Count_Top_Redis.scala
package test

import org.apache.flink.api.common.eventtime.WatermarkStrategy
import org.apache.flink.api.common.state.{ListState, ListStateDescriptor, ValueState, ValueStateDescriptor}
import org.apache.flink.connector.kafka.source.KafkaSource
import org.apache.flink.connector.kafka.source.enumerator.initializer.OffsetsInitializer
import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.functions.ProcessFunction
import org.apache.flink.streaming.api.functions.timestamps.BoundedOutOfOrdernessTimestampExtractor
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.scala.function.WindowFunction
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.streaming.api.windowing.triggers.EventTimeTrigger
import org.apache.flink.streaming.api.windowing.windows.TimeWindow
import org.apache.flink.streaming.connectors.redis.common.config.FlinkJedisPoolConfig
import org.apache.flink.streaming.connectors.redis.common.mapper.{RedisCommand, RedisCommandDescription, RedisMapper}
import org.apache.flink.streaming.util.serialization.SimpleStringSchema
import org.apache.flink.util.Collector
import org.json4s.JsonAST.{JDecimal, JInt, JString}
import org.json4s.jackson.JsonMethods

import java.text.SimpleDateFormat
import scala.jdk.CollectionConverters.{iterableAsScalaIterableConverter, seqAsJavaListConverter}

object Count_Top_Redis {

  private val Bigdata1 = "192.168.45.16"
  private val BootstrapServers = s"$Bigdata1:9092"
  private val topic = "ods_mall_data"
  val MaxOutOfOrderness: Long = 5 * 1000L
  // "已发货", "已付款", "已下单","已签收",
  val tuple = Array("已发货", "已签收")
  val Amount = new OutputTag[(String,String)]("Amount")
  val Consumption = new OutputTag[(String,String)]("Consumption")

  case class tagdata(name: String, data: String)

  case class Order_master_data(
                                order_id: Long,
                                //                                order_sn: Long,
                                //                                customer_id: Long,
                                //                                shipping_user: String,
                                //                                province: String,
                                //                                city: String,
                                //                                address: String,
                                //                                order_source: String,
                                //                                payment_method: Int,
                                //                                order_money: Double,
                                //                                district_money: Double,
                                //                                shipping_money: Double,
                                payment_money: Double,
                                //                                shipping_comp_name: String,
                                //                                shipping_sn: Long,
                                create_time: Long,
                                //                                shipping_time: Long,
                                //                                pay_time: Long,
                                //                                receive_time: Long,
                                order_status: String,
                                //                                order_point: Int,
                                //                                invoice_title: String,
                                modified_time: Long
                              )

  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)
    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
    val kafkaSource = KafkaSource.builder[String]
      .setBootstrapServers(BootstrapServers)
      .setTopics(topic)
      .setStartingOffsets(OffsetsInitializer.earliest())
      .setValueOnlyDeserializer(new SimpleStringSchema())
      .build()
    val timeFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
    val orderStream = env.fromSource(kafkaSource, WatermarkStrategy.noWatermarks(), topic)
      .filter(_.contains("order_master"))
      .map(line => {
        val str = JsonMethods.parse(line, useBigDecimalForDouble = true)
        val JInt(order_id) = str \\ "order_id"
        val JDecimal(payment_money) = str \\ "payment_money"
        val JString(create_time) = str \\ "create_time"
        val JString(order_status) = str \\ "order_status"
        val JString(modified_time) = str \\ "modified_time"
        (Order_master_data(order_id = order_id.toLong, payment_money = payment_money.toDouble, create_time = create_time.toLong, order_status = order_status, modified_time = timeFormat.parse(modified_time).getTime),
          timeFormat.parse(modified_time).getTime)

      })
      .assignTimestampsAndWatermarks(new choseEventTime)
      .keyBy(_ => true)
      .window(TumblingEventTimeWindows.of(Time.minutes(1)))
      .trigger(EventTimeTrigger.create())
      .apply(new putEventData)
.keyBy(_=>true)
      .process(new myPF)
    val config = new FlinkJedisPoolConfig.Builder()
      .setHost(Bigdata1)
      .setPort(6379)
      .build()
    orderStream.getSideOutput(Amount).print()
    orderStream.getSideOutput(Consumption).print()
    orderStream.print()
    //    orderStream.getSideOutput(Amount).addSink(new RedisSink[(String, String)](config, new redisSInk))
    //    orderStream.getSideOutput(Consumption).addSink(new RedisSink[(String, String)](config, new redisSInk))
    //    orderStream.addSink(new RedisSink[(String, String)](config, new redisSInk))
    env.execute()
  }

  class myPF extends ProcessFunction[Order_master_data, (String, String)] {
    lazy val totalCount: ValueState[Int] = getRuntimeContext.getState(new ValueStateDescriptor[Int]("totalCount", classOf[Int]))
    lazy val top3ItemAmount: ListState[(Long, Int)] = getRuntimeContext.getListState(new ListStateDescriptor[(Long, Int)]("top3ItemAmount", classOf[(Long, Int)]))
    lazy val top3ItemConsumption: ListState[(Long, Double)] = getRuntimeContext.getListState(new ListStateDescriptor[(Long, Double)]("top3ItemConsumption", classOf[(Long, Double)]))

    override def processElement(i: Order_master_data, context: ProcessFunction[Order_master_data, (String, String)]#Context, collector: Collector[(String, String)]): Unit = {
      if (tuple.contains(i.order_status)) {
        totalCount.update(totalCount.value() + 1)
        collector.collect("totalcount", totalCount.value().toString)

        top3ItemAmount.add(i.order_id, 1)
        val AmountList = top3ItemAmount.get().asScala.groupBy(_._1)
          .mapValues(_.map(_._2).sum)
          .toList
          .sortBy(_._2).reverse
        top3ItemAmount.clear()
        top3ItemAmount.addAll(AmountList.asJava)
        val put1 = AmountList :+ "null" :+ "null"
        context.output(Amount, ("top3itemamount", s"${put1(0)},${put1(1)}${put1(2)}"))
//        collector.collect("top3itemamount", s"${put1(0)},${put1(1)}${put1(2)}")

        top3ItemConsumption.add(i.order_id, i.payment_money)
        val ConsumptionList = top3ItemConsumption.get().asScala.groupBy(_._1)
          .mapValues(_.map(_._2).sum)
          .toList
          .sortBy(_._2).reverse
        top3ItemConsumption.clear()
        top3ItemConsumption.addAll(ConsumptionList.asJava)
        val put2 = ConsumptionList :+ "null" :+ "null"
        context.output(Consumption, ("top3itemconsumption", s"${put2(0)}${put2(1)}${put2(2)}"))
//        collector.collect("top3itemconsumption", s"${put2(0)}${put2(1)}${put2(2)}")
      }
    }
  }

  class choseEventTime extends BoundedOutOfOrdernessTimestampExtractor[(Order_master_data,Long)](Time.seconds(5)) {
    override def extractTimestamp(t: (Order_master_data,Long)): Long = t._2
  }

  class putEventData extends WindowFunction[(Order_master_data,Long), Order_master_data, Boolean, TimeWindow] {
    override def apply(key: Boolean, window: TimeWindow, input: Iterable[(Order_master_data,Long)], out: Collector[Order_master_data]): Unit = {
      // 对窗口内的数据按时间戳进行排序
      val sortedInput = input.toList
      // 发出按时间顺序排序后的数据
      sortedInput.foreach { case (line, timestamp) =>
        out.collect(line)
      }
    }
  }

  class redisSInk extends RedisMapper[(String, String)] {
    override def getCommandDescription: RedisCommandDescription = new RedisCommandDescription(RedisCommand.SET)

    override def getKeyFromData(t: (String, String)): String = t._1

    override def getValueFromData(t: (String, String)): String = t._2
  }
}

File: Task/src/main/java/org/example/gs6/countTopRedis.java
package org.example.gs6;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.state.ValueState;
import org.apache.flink.api.common.state.ValueStateDescriptor;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.connector.kafka.source.KafkaSource;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.ProcessFunction;
import org.apache.flink.util.Collector;
import org.apache.flink.util.OutputTag;
import org.example.*;

import java.text.SimpleDateFormat;
import java.util.Arrays;

public class countTopRedis implements Constants {
    public static void main(String[] args) throws Exception {
        String topic = "ods_mall_data";
        SimpleDateFormat timeFormat = new SimpleDateFormat("yyyyMMddHHmmss");
        OutputTag<String> top3itemamountlist = new OutputTag<String>("top3itemamountlist") {
        };
        OutputTag<Tuple2<String, Double>[]> top3itemconsumptionlist = new OutputTag<Tuple2<String, Double>[]>("top3itemconsumptionlist") {
        };
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
        String[] cleans = new String[]{"top3itemamountlist", "top3itemconsumptionlist"};
        ProcessFunction<OrderMaster, String> processFunction = new ProcessFunction<OrderMaster, String>() {
            private transient ValueState<Integer> priceCount;

            @Override
            public void open(Configuration parameters) throws Exception {
                priceCount = getRuntimeContext().getState(new ValueStateDescriptor<Integer>("PriceCount", Integer.class));
            }

            @Override
            public void processElement(OrderMaster ordermaster, ProcessFunction<OrderMaster, String>.Context context, Collector<String> collector) throws Exception {
                if (Arrays.asList(filter).contains(ordermaster.getOrder_status())) {
                    Integer value = priceCount.value();
                    if (value == null) value = 0;
                    int tmp = value + 1;
                    priceCount.update(tmp);
                    collector.collect(priceCount.value().toString());
                    TopN topN = new TopN();
                    topN.addElem("top3itemamountlist", 1, ordermaster.getOrder_id() + "");
                    Tuple2<String, Double>[] top3itemamountliststmp = topN.getTopN("top3itemamountlist", 0, 3);
                    Tuple2<String, Integer>[] top3itemamountlists = new Tuple2[top3itemamountliststmp.length];
                    for (int i = 0; i < top3itemamountliststmp.length; i++) {
                        top3itemamountlists[i] = new Tuple2(top3itemamountliststmp[i].f0, top3itemamountliststmp[i].f1);
                    }
                    String s = "";
                    for (int i = 0; i < top3itemamountlists.length; i++) {
                        String s1 = "";
                        if (i < top3itemamountlists.length-1) s1 = ",";
                        s = s + top3itemamountlists[i].f0 + ":" + top3itemamountlists[i].f1 + s1;
                    }
                    String tupleTopToString = "top3itemamount:[" + s + "]";
                    context.output(top3itemamountlist, tupleTopToString);
                    topN.addElem("top3itemconsumptionlist", ordermaster.getPayment_money(), ordermaster.getOrder_id() + "");
                    Tuple2<String, Double>[] top3itemconsumptionlists = topN.getTopN("top3itemconsumptionlist", 0, 3);
                    context.output(top3itemconsumptionlist, top3itemconsumptionlists);
                }
            }
        };
        KafkaSource<String> source = new KafkaConnect().kafkaReader(topic);

        WatermarkStrategy<OrderMaster> orderMasterWatermarkStrategy = new Watermarks().orderMasterWatermarkStrategy();
        SingleOutputStreamOperator<OdsMallData> odsMallDataStream = env
//                .fromSource(source, WatermarkStrategy.noWatermarks(), topic)
                .readTextFile("E:\\input\\order_master.txt")
                .map(in -> {
                    Gson gson = new GsonBuilder()
                            .create();
                    OdsMallData odsMallData = gson.fromJson(in, OdsMallData.class);
                    return odsMallData;
                });
        SingleOutputStreamOperator<OrderMaster> orderMasterStream = odsMallDataStream
                .filter(i -> i.getTable().equals("order_master"))
                .map(i -> {
                    Gson gson = new Gson();
                    String json = gson.toJson(i.getData());
                    OrderMaster orderMaster = gson.fromJson(json, OrderMaster.class);
                    return orderMaster;
                })
                .assignTimestampsAndWatermarks(orderMasterWatermarkStrategy);

        SingleOutputStreamOperator<String> data = orderMasterStream
                .keyBy(order_master -> true)
                .process(processFunction);
//        data.addSink(new RedisWriter().redisSet("totalprice"));
        data.print();
//        orderMasterStream.print();
        DataStream<String> top3itemamount = data.getSideOutput(top3itemamountlist);
//        top3itemamount.addSink(new RedisWriter().redisSet("top3itemamount"));
        top3itemamount.print();

        SingleOutputStreamOperator<String> top3itemconsumption = data.getSideOutput(top3itemconsumptionlist).map(tuple2s -> {
            String s = "";
            for (int i = 0; i < tuple2s.length; i++) {
                String s1 = "";
                if (i < tuple2s.length-1) s1 = ",";
                s = s + tuple2s[i].f0 + ":" + tuple2s[i].f1 + s1;
            }
            return "top3itemconsumption:[" + s + "]";
        });
//        top3itemconsumption.addSink(new RedisWriter().redisSet("top3itemconsumption"));

        env.execute("countRedisMysql");
    }
}
